org-fc-position.el

```diff
@@ -0,0 +1,80 @@
+;;; org-fc-position.el --- Representation of a position -*- lexical-binding: t; -*-
+;;
+;; Copyright (C) 2022 Cash Weaver
+;;
+;; Author: Cash Weaver <cashbweaver@gmail.com>
+;; Maintainer: Cash Weaver <cashbweaver@gmail.com>
+;; Created: September 26, 2022
+;; Modified: September 26, 2022
+;; Version: 0.0.1
+;; Keywords: abbrev bib c calendar comm convenience data docs emulations extensions faces files frames games hardware help hypermedia i18n internal languages lisp local maint mail matching mouse multimedia news outlines processes terminals tex tools unix vc wp
+;; Homepage: https://github.com/cashweaver/org-fc-position
+;; Package-Requires: ((emacs "24.3"))
+;;
+;; This file is not part of GNU Emacs.
+;;
+;;; Commentary:
+;;
+;; A position is a single "view" of a card.
+;;
+;;; Code:
+
+(require 'eieio)
+(require 'dash)
+
+(defclass org-fc-position ()
+  ((box
+    :initform 0
+    :initarg :box
+    :type (integer -1 *)
+    :documentation "The positions's current Leitner system box.")
+   (card
+    :initform nil
+    :initarg :card
+    :documentation "The `org-fc-card' which contains this position.")
+   (due
+    :initform nil
+    :initarg :due
+    :type list
+    :custom (repeat integer)
+    :documentation "The time at which this position is due.")
+   (ease
+    :initform 0
+    :initarg :ease
+    :type (or float integer)
+    :documentation "The ease factor.")
+   (interval
+    :initform 0
+    :initarg :interval
+    :type (or float integer)
+    :documentation "The interval, in days.")
+   (pos
+    :initform ""
+    :initarg :pos
+    :type (or integer string)
+    :documentation "The name of the position (e.g. \"front\" for double/normal or \"0\" for cloze)."))
+  "Represents a single position.")
+
+(cl-defmethod org-fc-position--is-due ((pos org-fc-position))
+  "Return t if POS is due; else nil."
+  (time-less-p (oref pos due) (current-time)))
+
+(defun org-fc-positions--filter-due (positions)
+  "Filter POSITIONS to include only enabled and due positions."
+  (let ((due-enabled-positions (cl-loop for pos in positions
+                                        when (and (not (oref (oref pos card) suspended))
+                                                  (org-fc-position--is-due pos))
+                                        collect pos)))
+    (if org-fc-bury-siblings
+        (org-fc-positions--one-per-card due-enabled-positions)
+      due-enabled-positions)))
+
+(defun org-fc-positions--one-per-card (positions)
+  "Return filtered list of POSITIONS; only one position per card."
+  (let ((-compare-fn (lambda (a b)
+                       (equal (oref (oref a card) id)
+                              (oref (oref b card) id)))))
+    (-distinct positions)))
```

Comment:
I found the use of `-distinct` to be more legible than the alternative. I'm open to suggestions as this uses `dash` in contrast to the existing implementation.

Reply:
I've been thinking about this, my original motivation was to avoid adding an dependency
but if you'd like to use dash.el I'm ok with that.

Another question is whether we want to use the anaphoric functions, too.
Initially that would make it harder to understand the code but the same argument can be made
against the use of `cl-loop`. It should be fine if we note it somewhere in the documentation.

Reply:
> Another question is whether we want to use the anaphoric functions, too.

I don't think there is a anaphoric counterpart for `-distinct`/`-uniq`.

> [...] the same argument can be made against the use of `cl-loop`.

Would you like me to swap the `cl-loop` out for a `--filter`?

Reply:
Please don't take what I wrote as a critique of your code.

The point I'm trying to make is that the increase in expressiveness of any library or complex macro (`cl-loop`, anaphoric dash.el) comes at the cost of an initial hurdle to understanding the code for new developers.

That's ok as long as there's kind of “before jumping into the code, check out these docs” disclaimer,
and as long as we're in agreement on which (all?) features we'll use.

I'm fine with adding dash and using all the features it provides,
e.g. avoiding “wrapper-lambdas” when possible.

Not sure about how existing code should be handled,
refactoring it just to fit the newer style seems like a lot of effort
but we can just fix it as we come across it.